### DISCUSSION TITLE
_createSuper loses context when Reflect.Construct is available in new version of Babel

### DISCUSSION THREAD

**Discussion by OnlineInternetPerson** [babel/babel#14103] [Q&A] [answered]
I'm hoping someone might have the fix to an issue I'm running into when upgrading babel to a new version. I'm posting here because babel is producing code I don't expect. I'm running babel as a browserify transform (babelify). A little bit at wit's end with this one. I'm starting to think this is either a babel bug or a use case that needs better documentation. Any help would be appreciated!
                             
Here are the versions:

              Old version:
                             babel-core: ^6.26.x
                             babel-preset-env: ^1.7.x
                             babelify: ^8.0.x
                             browserify: ^7.0.x

              New versions:
                             @babel/core: ^7.16.x
                             @babel/preset-env: ^7.16.x
                             babelify: ^10.0.x
                             browserify: ^17.0.x

                             
My source class:

              class MyClass extends MySuperClass {
                             super(options);
              }
              someMethod() {
                             return someOtherThing;
              }
              export default MyClass;

Old compiled Babel class (good!):

              function MyClass(options) {
                             _classCallCheck(this, MyClass);
                             <snip, who cares>
                             // Note: The MySuperClass constructor is availabe on MyClass.__proto__ on the line below
                             return _possibleConstructorReturn(this, (MyClass.__proto__ || Object.getPrototypeOf(MyClass)).call(this, options));
              }

New compiled Babel class (not working):

    var MyClass = /*#__PURE__*/
        function(_MySuperClass) {
            _inherits(MyClass, _MySuperClass);
            var _super = _createSuper(MyClass);
            function MyClass(options) {
                _classCallCheck(this, MyClass);
                <snip, who cares>
                return _super.call(this, options);
            }
         }

I've traced the problem is to the _createSuper method, which loses the context of the class when it calls Reflect.construct.
As an example, there's this github test which points to the sort of code it's using:
babel/packages/babel-plugin-transform-runtime/test/fixtures/use-options/useESModules-cjs-auto/options.json

    function _createSuper(Derived) {
        var hasNativeReflectConstruct = _isNativeReflectConstruct();
                             return function _createSuperInternal() {
            var Super = _getPrototypeOf(Derived), result;
            if (hasNativeReflectConstruct) {
                var NewTarget = _getPrototypeOf(this).constructor;
                                                          // NOTE: correct context is available in the scope as "this", but 
                                                          // instead a new context is created by Reflect.construct using the NewTarget
                                                          // (which is the correct constructor, but we need the same context)
                result = Reflect.construct(Super, arguments, NewTarget);
            } else {
                result = Super.apply(this, arguments);
            }
            return _possibleConstructorReturn(this, result);
        }
        ;
    }
                             
One interesting point is that if I put a breakpoint _isNativeReflectConstruct and set Reflect to undefined, the code works perfectly. This is why I am suspicious of a babel bug. Any help would be appreciated!

I noticed adding this seems to help, but causes other compilation issues. "@babel/plugin-transform-classes", { loose: true }

**Comment by nicolo-ribaudo** [comment, marked as answer]
How are you using that class?

**Reply by OnlineInternetPerson** [reply]
Using CommonJS and ES5:

    var MyClass = require("./path/to/MyClass").default

    function MyCustomSubclass() {
      MyClass.apply(this, arguments);
      this.property = "value";
    }

    MyCustomSubclass.prototype = Object.create(MyClass.prototype);
    MyCustomSubclass.prototype.constructor = MyCustomSubclass;

    MyCustomSubclass.prototype.foo = function() { <snip> }

**Reply by nicolo-ribaudo** [reply]
Ok, the problem is that `MyClass.apply` is not supposed to work with ES6 classes (you can test it running the untranspiled code directly). You should do this instead:

```js
function MyCustomSubclass() {
  var self = Reflect.construct(MyClass, arguments, new.target);
  self.property = "value";
  return self.
}
```

The reason is that `this` is created in the constructor upper class in the inheritance chain (in `MySuperClass`) in your case, and not in the lower one (`MyCustomSubclass`).

**Reply by OnlineInternetPerson** [reply]
Thanks for looking into this! I'm hesitant to chalk this up to ES6 class incompatibility ES5 inheritance patterns. One expectation is that babel is supposed to be compiling these classes into ES5-compatable syntax thanks to `@babel/preset-env`. Is this not accurate?

Looking at this block from the docs:

    https://babeljs.io/docs/en/babel-preset-env#no-targets
    Since one of the original goals of preset-env was to help users easily transition from using preset-latest, 
    it behaves similarly when no  targets are specified: preset-env will transform all ES2015-ES2020 code to 
    be ES5 compatible.

**Reply by nicolo-ribaudo** [reply]
Babel's goal is to produce code that can run in ES5 environments and that works as closely as possible to how ES6 works. It's not to transform ES6 idioms into ES5 idioms.

Unfortunately ES6 classes are not compatible with the inheritance mechanism that was used in ES5, but they need to be inherited either using the `class ... extends` syntax or as I shown in https://github.com/babel/babel/discussions/14103#discussioncomment-1914572
